PROTOCOL.md

```diff
@@ -18,52 +21,54 @@ export interface OperationMessage {
-    * `context?: any` : Execution context for the operation
     
 #### GQL_STOP
 Client sends this message in order to stop a running GraphQL operation execution (for example: unsubscribe)
 - `id: string` : operation id
     
+#### GQL_CONNECTION_TERMINATE
+Client sends this message to terminate the connection.    
+    
 ### Server -> Client
 
 #### GQL_CONNECTION_ERROR
-The server sends this message if `onConnect` callback returns `false` or throws an exception, and after sending this message, the server disconnects the client.
+The server may responses with this message to the `GQL_CONNECTION_INIT` from client, indicates the server rejected the connection.
 
 It server also respond with this message in case of a parsing errors of the message (which does not disconnect the client, just ignore the message).
 
 - `payload: Object`: the server side error
 
 #### GQL_CONNECTION_ACK
-The server sends this message if `onConnect` callback returns any other value then `false`.
+The server may responses with this message to the `GQL_CONNECTION_INIT` from client, indicates the server accepted the connection.
+
+The server will response only with `GQL_CONNECTION_ACK` or `GQL_CONNECTION_ERROR` to `GQL_CONNECTION_INIT` message.
 
 #### GQL_DATA
-The server sends this message to transfter the GraphQL execution result from the server to the client.
+The server sends this message to transfter the GraphQL execution result from the server to the client, this message is a response for `GQL_START` message.
+
+For each GraphQL operation send with `GQL_START`, the server will respond with at least one `GQL_DATA` message.
+
 - `id: string` : ID of the operation that was successfully set up
 - `payload: Object` : 
     * `data: any`: Execution result
     * `errors?: Error[]` : Array of resolvers errors
 
 #### GQL_ERROR
-Server sends this message upon failing to **register** or **execute** a GraphQL operation (resolver errors are part of `GQL_DATA` message, and will be added as `errors` array)
-- `payload: { errors: Array<Object> }` : payload with an array of errors attributed to the subscription failing on the server
+Server sends this message upon a failing operation, before the GraphQL execution, usually due to GraphQL validation errors (resolver errors are part of `GQL_DATA` message, and will be added as `errors` array)
+- `payload: Error` : payload with the error attributed to the operation failing on the server
 - `id: string` : operation ID of the operation that failed on the server
 
 #### GQL_COMPLETE
-Server sends this message to indicate that a GraphQL operation is done, and no more data will arrive (for queries and subscription, it called immediatly, and for subscriptions it called when unsubscribing)
+Server sends this message to indicate that a GraphQL operation is done, and no more data will arrive (for queries and mutations, it called immediatly, and for subscriptions it called when unsubscribing)
```

Comment:
`for queries and mutation, it called immediately` that's not accurate as this will make the implemantation of reactive graphql invalid.